storage/store.go

```diff
@@ -1510,16 +1518,21 @@ func (s *Store) ProposeRaftCommand(idKey cmdIDKey, cmd proto.RaftCommand) <-chan
 	if err != nil {
 		log.Fatal(err)
 	}
-	etr, ok := value.(*proto.EndTransactionRequest)
-	if ok && etr.InternalCommitTrigger != nil &&
-		etr.InternalCommitTrigger.ChangeReplicasTrigger != nil {
-		// EndTransactionRequest with a ChangeReplicasTrigger is special because raft
-		// needs to understand it; it cannot simply be an opaque command.
-		crt := etr.InternalCommitTrigger.ChangeReplicasTrigger
-		return s.multiraft.ChangeGroupMembership(cmd.RangeID, string(idKey),
-			changeTypeInternalToRaft[crt.ChangeType],
-			proto.MakeRaftNodeID(crt.NodeID, crt.StoreID),
-			data)
+	for _, union := range cmd.Cmd.Requests {
+		args := union.GetValue().(proto.Request)
+		etr, ok := args.(*proto.EndTransactionRequest)
+		if ok && etr.InternalCommitTrigger != nil && etr.InternalCommitTrigger.ChangeReplicasTrigger != nil {
+			if len(cmd.Cmd.Requests) != 1 {
+				panic("EndTransaction should only ever occur by itself in a batch")
```

Comment:
s/EndTransaction/EndTransaction with ChangeReplicasTrigger/

This check may be too harsh. There can only be one such request in the batch, and I don't think there's any problem with combining other parts of the transaction in the final batch.